### ISSUE TITLE
rule to detect unnecessary `is True`/`is False`

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#6543]
```py
def foo(value: bool):
    if value is True: # should be `if value`
        ...
    if value is False: # should be: `if not value`
        ...
```

**Comment by heidi666** [comment]
These aren't equivalent though, unless you're taking the type hint into account when you're evaluating the rule.

**Comment by DetachHead** [comment]
yeah it should take the type into account

**Comment by NeilGirdhar** [comment]
> These aren't equivalent though, unless you're taking the type hint into account when you're evaluating the rule.

You're right that they're not equivalent.

I'm not sure if you're arguing against there being a rule, but I think that there should be a rule.  If you want to check the type and its value, then I prefer:
```python
if isinstance(value, bool) and value: ...
```
or
```python
match value:
    case True: ...
```
These constructs (`isinstance` and `match`) make it clear that switching on type is happening.

**Comment by DetachHead** [comment]
i think #3893 should be prioritized, since it would make implementing rules like this easier

**Comment by heidi666** [comment]
@NeilGirdhar 
> I think that there should be a rule irrespective of type inference.

I disagree. Checking with `is` is the commonly used and recommended way (see [PEP8](https://peps.python.org/pep-0008/#programming-recommendations)) to check for singletons like None, True, False etc. Assuming that you actually do want to check for those values and not general truthiness.

**Comment by NeilGirdhar** [comment]
> I disagree. Checking with `is` is the commonly used and recommended way (see [PEP8](https://peps.python.org/pep-0008/#programming-recommendations)) to check for singletons like None, True, False etc. As

Yes, it's recommended for `None`.  Is there anyone recommends it for `True` and `False`?  The [Google Python](https://google.github.io/styleguide/pyguide.html) style guide forbids using `==`, and explicitly recommends `if x`.

I realize that this may be opinionated, so do you have any examples in code where `if x is True` is beneficial?

It seems unnecessarily wordy (in the case of a `bool`) or an implicit type switch, which seems like ugly code.  I see your point that they're singletons, but they're also values of the type `bool`, so this comparison is an implicit type switch, which leads to weird code:
```python
if x is True:
  ...
elif x is False:
  ...
else:
  ... # The type checker won't deduce that x is not Boolean here.
```
versus:
```
if isinstance(x, bool):
  if x:
    ...
  else:
    ...
else:
  ... # the type checker knows that x is Boolean here.
```
I think the latter code is much clearer.
